contracts/protocol/libraries/logic/PoolExtendedLogic.sol

```diff
@@ -0,0 +1,367 @@
+            request.borrowAmount,
+            request.collection,
+            request.tokenId
+        );
+    }
+
+    function executeRejectBlurExchangeRequest(
+        DataTypes.PoolStorage storage ps,
+        IPoolAddressesProvider poolAddressProvider,
+        DataTypes.BlurBuyWithCreditRequest calldata request
+    ) external {
+        // check request status
+        bytes32 requestHash = _calculateBlurExchangeRequestHash(request);
+        require(
+            ps._blurExchangeRequestStatus[requestHash] ==
+                DataTypes.BlurBuyWithCreditRequestStatus.Initiated,
+            Errors.INVALID_REQUEST_STATUS
+        );
+
+        address keeper = ps._blurExchangeKeeper;
+        require(msg.sender == keeper, Errors.CALLER_NOT_KEEPER);
+
+        ps._blurOngoingRequestAmount -= 1;
+
+        //repay and supply weth for user
+        address weth = poolAddressProvider.getWETH();
+        require(msg.value == request.listingPrice, Errors.INVALID_ETH_VALUE);
+        IWETH(weth).deposit{value: request.listingPrice}();
+        repayAndSupplyForUser(
+            ps,
+            weth,
+            address(this),
+            request.initiator,
+            request.listingPrice
+        );
+
+        //burn nToken.
+        uint256[] memory tokenIds = new uint256[](1);
+        tokenIds[0] = request.tokenId;
+        burnUserNToken(
+            ps,
+            request.collection,
+            tokenIds,
+            false,
+            request.initiator
+        );
+
+        delete ps._blurExchangeRequestStatus[requestHash];
+
+        emit BlurExchangeRequestRejected(
+            request.initiator,
+            request.paymentToken,
+            request.listingPrice,
+            request.borrowAmount,
+            request.collection,
+            request.tokenId
+        );
+    }
+
+    function _calculateBlurExchangeRequestHash(
+        DataTypes.BlurBuyWithCreditRequest calldata request
+    ) internal pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encode(
+                    request.initiator,
+                    request.paymentToken,
+                    request.listingPrice,
+                    request.borrowAmount,
+                    request.collection,
+                    request.tokenId
+                )
+            );
+    }
+
+    function repayAndSupplyForUser(
+        DataTypes.PoolStorage storage ps,
+        address asset,
+        address payer,
+        address onBehalfOf,
+        uint256 totalAmount
+    ) public {
+        address variableDebtTokenAddress = ps
+            ._reserves[asset]
+            .variableDebtTokenAddress;
+        uint256 repayAmount = Math.min(
+            IERC20(variableDebtTokenAddress).balanceOf(onBehalfOf),
+            totalAmount
+        );
+        repayForUser(ps, asset, payer, onBehalfOf, repayAmount);
+        supplyForUser(ps, asset, payer, onBehalfOf, totalAmount - repayAmount);
+    }
+
+    function supplyForUser(
```

Comment:
Should we do two logic files? one for the extended generic logic  and one for blur (or add the blur in the marketplace logic)